Shell scripts to calculate smart contract sizes

arthurgousset opened the pull request:
### Description

Sometimes it's useful to know the bytecode size of core contracts. This PR adds scripts calculate that in two ways:

1. Using the bytecode of actually deployed smart contracts.
2. Using the build artifacts of locally built smart contracts.

Writes the output to the console and to csv files in the `scripts/bash/out/` directory (which is gitignored).

### Changes

1. Adds two shell scripts
2. Adds commands to `package.json`
3. Adds documentation to `README.md`

### Other changes

Nit: Linting in `README.md` file

### Tested

- [x] Example run `size:onchain`

    ```sh
    $ yarn size:onchain
    
    # ...
    StableTokenEUR,0x434563B0604BE100F04B7Ae485BcafE3c9D8850E,9.180
    Validators,0xe52EaC18fB3C1e1713e73d4A5b7dCb12a2f2C697,58.228
    Data extraction complete. Results saved to /Users/arthur/Documents/celo-org/celo-monorepo/packages/protocol/out/onchain_bytecode_sizes_20240509_145556.csv
    ✨  Done in 19.15s.
    ```

- [x] Example run `size:artifacts`

    ```sh
    $ yarn size:artifacts
    
    # ...
    ReleaseGold,31.721
    OdisPaymentsProxy,2.868
    Data extraction complete. Results saved to /Users/arthur/Documents/celo-org/celo-monorepo/packages/protocol/scripts/bash/out/build_artefact_bytecode_sizes_20240509_151111.csv
    ✨  Done in 9.57s.
    ```

- [x] CSV output for both scripts looks good 👍 
- [x] Output in `out/` is gitignored

### Related issues

- Fixes #10988 

### Backwards compatibility

Yes, backwards compatible. The new scripts can be used on-demand, but don't interfere with any existing flows. 

### Documentation

Yes, added documentation to the `README.md` file.

arthurgousset committed:
chore(scripts): adds two shell scripts

1. to calculate smart contract sizes from actual deployment addresses
2. to calculate smart contact sizes from local build artefacts

arthurgousset committed:
style(protocol/README): linting (spaces, new lines, indentation)

arthurgousset committed:
chore(scripts): renames file to fix typo "artefacts" -> "artifacts"

arthurgousset committed:
fix(scripts): find build artifacts using relative directories

arthurgousset committed:
chore(protocol): adds two `size` commands in `package.json`

arthurgousset committed:
docs(protocol/README): adds documentation for two scripts

arthurgousset committed:
fix(scripts): modifies output directory

soloseng reviewed (approved):
🚀

m-chrzan commented on packages/protocol/scripts/bash/get_smart_contract_size_from_build_artifacts.sh:
```diff
@@ -0,0 +1,68 @@
+do
+    # Extract the contract name and bytecode using jq
+    contract=$(jq -r '.contractName' "$file")
+    bytecode=$(jq -r '.bytecode' "$file")
```
Should we be getting the `.deployedBytecode` instead? The EIP-170 limit refers to the deployed bytecode, not the initialization bytecode.

arthurgousset replied:
That's a great point, [EIP-170](https://eips.ethereum.org/EIPS/eip-170) puts "a hard cap on the size of an object that can be **saved to the blockchain**".

Since it's the `deployedBytecode` that will be stored in the smart contract, and not the initialization `bytecode`, we should be looking at the size of the `deployedBytecode`.

That might also explain (parts of) why the on-chain calculation is slightly different, since that uses the `eth_getCode` RPC method, which returns the deployed bytecode:

> The command `eth_getCode` returns the bytecode associated with the address in the Ethereum world state, that will be the deployed bytecode (or runtime bytecode).
> 
> The init bytecode could be obtained from the deployment transaction's data field, unless it was created by a contract in which case you have to trace the transaction to obtain the details.

Source: [ethereum.stackexchange.com](https://ethereum.stackexchange.com/questions/108152/does-eth-getcode-return-the-init-bytecode-or-the-deployed-bytecode)

Let me update this 👍

arthurgousset replied:
That's done in https://github.com/celo-org/celo-monorepo/pull/10989/commits/c128c7f1b417a8cd82fab2d93c84c4ea00b50678. 

With this change, the contract sizes using the on-chain method and the build artifact method are equal.

<img width="1044" alt="image" src="https://github.com/celo-org/celo-monorepo/assets/46296830/2383fadf-54ef-47de-ba64-f360eb43544e">

Source: [Contract sizes](https://docs.google.com/spreadsheets/d/1MPo5mtC8DT8d3y1ovGGhj04rQ3Zx6Xb0prgz7F4VNQM/edit#gid=1494083158)

arthurgousset committed:
fix(scripts:artifacts): uses `deployedBytecode` instead of `bytecode`

[EIP-170](https://eips.ethereum.org/EIPS/eip-170) puts "a hard cap on the size of an object that can be saved to the blockchain".

Since it's the deployedBytecode that will be stored in the smart contract, and not the initialization bytecode, we should be looking at the size of the deployedBytecode.

That might also explain (parts of) why the on-chain calculation is slightly different, since that uses the eth_getCode RPC method, which returns the deployed bytecode.

With this change the contract size using on-chain method and the build artifact method are equal.

arthurgousset committed:
docs(protocol/README): replaces `bytecode` -> `deployedBytecode`

pahor167 commented on packages/protocol/scripts/bash/get_smart_contract_size_from_onchain_address.sh:
```diff
@@ -0,0 +1,83 @@
+TEMP_FILE="temp_$TIMESTAMP.txt"
+
+# Set the RPC URL for the blockchain
+RPC_URL="https://forno.celo.org"
```
We might consider also supporting alfajores

arthurgousset replied:
Sounds good, I added support for custom RPC URLs (e.g. `https://alfajores-forno.celo-testnet.org`) in https://github.com/celo-org/celo-monorepo/pull/10989/commits/7e5a6c6b8f52ceb8346aec75ea310408f5a3edac.

If no `RPC_URL` is given, it defaults to Mainnet.

I updated the `README` with instructions in https://github.com/celo-org/celo-monorepo/pull/10989/commits/1163dd97e709eef8eb24af012ab19a17122d0d79.

arthurgousset committed:
feat(scripts/size:onchain): adds support for custom RPC URL

arthurgousset committed:
docs(protocol/README): adds description to use custom RPC URL

arthurgousset committed:
Merge branch 'master' into arthurgousset/contract-size-shell-scripts
